tests/Unit/Core/Grid/Column/ColumnCollectionTest.php

```diff
@@ -0,0 +1,187 @@
+            ->add($this->createColumnMock('second'))
+            ->add($this->createColumnMock('third'))
+        ;
+
+        $this->assertEquals([
+            'first',
+            'second',
+            'third',
+        ], $this->getColumnPositions($columns));
+
+        return $columns;
+    }
+
+    /**
+     * @depends testItAddsColumnsToCollection
+     */
+    public function testItAddsColumnsBeforeGivenColumn(ColumnCollection $columns)
+    {
+        $columns
+            ->addBefore('first', $this->createColumnMock('before_first'))
+            ->addBefore('second', $this->createColumnMock('before_second'))
+            ->addBefore('third', $this->createColumnMock('before_third'))
+        ;
+
+        $this->assertEquals([
+            'before_first',
+            'first',
+            'before_second',
+            'second',
+            'before_third',
+            'third',
+        ], $this->getColumnPositions($columns));
+
+        return $columns;
+    }
+
+    /**
+     * @depends testItAddsColumnsBeforeGivenColumn
+     */
+    public function testItAddsColumnsAfterGivenColumn(ColumnCollection $columns)
+    {
+        $columns
+            ->addAfter('first', $this->createColumnMock('after_first'))
+            ->addAfter('second', $this->createColumnMock('after_second'))
+            ->addAfter('third', $this->createColumnMock('after_third'))
+        ;
+
+        $this->assertEquals([
+            'before_first',
+            'first',
+            'after_first',
+            'before_second',
+            'second',
+            'after_second',
+            'before_third',
+            'third',
+            'after_third'
+        ], $this->getColumnPositions($columns));
+
+        return $columns;
+    }
+
+    /**
+     * @depends testItAddsColumnsAfterGivenColumn
+     */
+    public function testItRemovesColumnById(ColumnCollection $columns)
+    {
+        $columns->remove('first');
+        $columns->remove('third');
+        $columns->remove('non_existing');
+
+        $this->assertCount(7, $columns);
+    }
+
+    public function testMixingMethodsProducesCollectionWithCorrectColumnPositions()
+    {
+        $columns = (new ColumnCollection())
+            ->add($this->createColumnMock('first'))
+            ->addAfter('first', $this->createColumnMock('to_be_removed'))
+            ->addBefore('first', $this->createColumnMock('before_first'))
+            ->addBefore('first', $this->createColumnMock('before_first_2'))
+            ->addAfter('first', $this->createColumnMock('after_first'))
+            ->addBefore('first', $this->createColumnMock('before_first_3'))
+            ->remove('non_existing')
+            ->addAfter('after_first', $this->createColumnMock('after_first_2'))
+            ->add($this->createColumnMock('second'))
+            ->remove('to_be_removed')
+            ->add($this->createColumnMock('third'))
+            ->addAfter('second', $this->createColumnMock('after_second'))
```

Comment:
Now, can you test with other than string key, maybe numeric for example =)

Reply:
ping @sarjon this is not finished here ^^

Reply:
done.